### PR TITLE
add addresses controller, use case and gateway to that calls addresses api

### PR DESCRIPTION
**Pull request by elena-vi**
Co-authored-by: maysakanoni <maysa@madetech.com>
Co-authored-by: Bhavesh-Patel <bhavesh.patel@madetech.com>

Also in this pull request:
- make test and build steps run concurrently 
- only deploy when pushing to main

Important changes to the workflow [here](https://github.com/City-of-Lincoln-Council/housing-repairs-online-api/pull/2/files#diff-d5a2cea578a0446aad66faf31bf1076ae94c9916b1a3374e342272a6300e8344). Previously the tests were not running, so I removed the -no-build flag because that wasn’t running the tests properly and I removed the verbosity as it was overly polluting the logs, example of a failure without the flag can be seen [here](https://github.com/City-of-Lincoln-Council/housing-repairs-online-api/runs/3953489096?check_suite_focus=true)

### CODE REVIEW

File: HousingRepairsOnlineApi.Tests/GatewaysTests/AddressGatewayTests.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                    "\"CityName\": \"New Meow City\", " +
+                    "\"AddressLine\": [\"123 Cute Street\"], " +
+                    "\"Type\": \"null\", " +
+                    "\"PostalCode\": \"M3 0W\"}]");
+            // Act
+            var data = await addressGateway.Search(Postcode);
+
+            // Assert
+            mockHttp.VerifyNoOutstandingExpectation();
+            Assert.Equal(1, data.Count());
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/GatewaysTests/AddressGatewayTests.cs:69]: As the GitHub warning below suggests not to use `Assert.Equal`, use `Assert.Single` instead.
```suggestion
            Assert.Single(data);
```

File: HousingRepairsOnlineApi.Tests/GatewaysTests/AddressGatewayTests.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+            mockHttp.Expect($"https://our-porxy-UH.api/address?postcode={Postcode}")
+                .WithHeaders("X-API-Key", "super secret")
+                .Respond(statusCode: (HttpStatusCode)503);
+            // Act
+            var data = await addressGateway.Search(Postcode);
+
+            // Assert
+            mockHttp.VerifyNoOutstandingExpectation();
+            Assert.Equal(0, data.Count());
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi.Tests/GatewaysTests/AddressGatewayTests.cs:87]: As the GitHub warning below suggests not to use `Assert.Equal`, use `Assert.Empty` instead.
```suggestion
            Assert.Empty(data);
```

File: HousingRepairsOnlineApi/Controllers/AddressesController.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+        public AddressesController(IRetrieveAddressesUseCase retrieveAddressesUseCase)
+        {
+            this.retrieveAddressesUseCase = retrieveAddressesUseCase;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> Addresses([FromQuery] string postcode)
+        {
+            // TODO: add tests ensuring that query params are passed to the usecase
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/Controllers/AddressesController.cs:21]: Does this comment suggest additional tests need to be written?

File: HousingRepairsOnlineApi/UseCases/RetrieveAddressesUseCase.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    PostCode = address.PostalCode
+                }
+                ));
+            }
+            return result;
+        }
+    }
+}
+
+// "uprn": null,
```
Comment (Bhavesh-Patel) [inline HousingRepairsOnlineApi/UseCases/RetrieveAddressesUseCase.cs:43]: Remove these comment lines - unless required(?)

File: HousingRepairsOnlineApi/Controllers/AddressesController.cs (hunk 3 above)
Comment (Bhavesh-Patel) [reply HousingRepairsOnlineApi/Controllers/AddressesController.cs]: I can see the comment has been removed now - will resolve.

**Review by Bhavesh-Patel — APPROVED** [review]

**Comment by elena-vi** [comment]
[Gather Address Information (MVP)](https://trello.com/c/nb8dDxfb/81-gather-address-information-mvp)

**Pull request merged**